### ISSUE TITLE
[nvmf_vfio_user] Handle multiple 4KiB mappings

### ISSUE THREAD

**Issue by sboeuf** [spdk/spdk#2101]
I'm playing with the NVME implementation for vfio-user through SPDK, and particularly the case where the device is placed behind a virtual IOMMU.
I run into an error where the NVME queue 1 can't be allocated properly because its size is 16KiB while the mappings the client provided to the server are 4KiB, meaning the code in `spdk/libvfio-user/lib/dma.c` finds that we need 4 regions to cover the whole thing.
Problem is, there's a strong assumption in `spdk/lib/nvmf/vfio_user.c` `map_one()` function that makes the handling of multiple regions impossible.
```c
ret = vfu_addr_to_sg(ctx, (void *)(uintptr_t)addr, len, sg, 1, prot);
```
should not limit the scatter/gather maximum to 1, and later on:
```c
ret = vfu_map_sg(ctx, sg, iov, 1, 0);
```
should rely on the number of regions returned by the previous function to map correctly, instead of assuming 1 again.

I guess there's also multiple places where to fix this issue with the unmap code path as well.

To summarize, the guest might be using 4K pages, leading the vIOMMU to submit multiple mappings in order to cover the full range of addresses which must be accessed, and I think this should be supported. I don't think we should expect the DMA mappings to refer to 2MiB pages.

**Comment by changpe1** [comment]
@sboeuf Do you use huge pages in the client? currently huge pages are required because we don't remap the client memory region in the unit of 4KiB.

**Comment by sboeuf** [comment]
@changpe1 
>Do you use huge pages in the client?

Even when I use 2MiB hugepages in the client, the vIOMMU asks for 4KiB mappings, which the client forwards to the vfio-user server. I don't think I can force the vIOMMU to issue 2MiB mappings only.

>currently huge pages are required because we don't remap the client memory region in the unit of 4KiB.

Are you planning to do so in the future? I'm asking because if this is about time, I can wait for this to be implemented.
